src/app/pages/products-list/card/card.component.ts

```diff
@@ -1,32 +1,34 @@
-import {Component} from '@angular/core';
-import {productsMock} from '../../../shared/products/products.mock';
-
-// const user: User = {
-
-// }
+import {Component, EventEmitter, Input, Output} from '@angular/core';
+import {IProduct} from '../../../shared/products/product.interface';
 
 @Component({
     selector: 'app-card',
     templateUrl: './card.component.html',
     styleUrls: ['./card.component.css'],
 })
 export class CardComponent {
-    readonly product = productsMock[0];
-    // user: {name: string} = {name: 'Alex'};
-    // user = user;
+    @Input({required: true}) product!: IProduct;
+    @Output() itemToCart = new EventEmitter<string>();
+    @Output() itemToFavorite = new EventEmitter<string>();
+    currentSLide = 0;
 
-    // constructor() {
-    //     this.user = null;
-    // }
-
-    onProductBuy(event: Event) {
-        event.stopPropagation();
+    buyProduct(productId: string): void {
+        this.itemToCart.emit(productId);
+    }
 
-        // eslint-disable-next-line no-console
-        console.log('Buy product');
+    buyFavorite(productId: string): void {
+        this.itemToFavorite.emit(productId);
     }
 
     isStarActive(starIndex: number): boolean {
         return this.product.rating >= starIndex;
     }
+
+    nextSlide(): void {
+        this.currentSLide = ++this.currentSLide % this.product.images.length;
```

Comment:
Я являюсь противником неявных изменений, как, например, тут `++this.currentSLide`: `++` в такой веренице кода теряется, то есть его нужно заметить и осознать, следовательно читаемость усложняется и код становится грязнее. 
В противовес этому я готов поставить явную инициализацию, которая проста в чтении и заметна
```suggestion
        this.currentSLide += 1;
        this.currentSLide = this.currentSLide % this.product.images.length;
```

Тут конечно возникает вопрос: а что мы хотим сделать? Потому что такие строки вводят в ступор - тут прибавили единицу, а тут перевычислили. По этому я рекомендую это вынести либо в отдельную функцию-утилиту с говорящим неймингом, чтобы другие разработчики не вникали в реализацию зацикливания, либо реализовать зацикливание галереи более явно. Но дестркутуризация здесь, я считаю, маст хэв. Вот такое тело функции я бы предложил:
```ts
        const nextSlide = currentSLide + 1;

        return nextSlide % this.product.images.length;
```

Reply:
В моем варианте используется префиксный инкремент ++, который сначала увеличивает значение this.currentSlide, а затем выполняет операцию взятия остатка от деления на длину массива изображений. Это обеспечивает, что this.currentSlide всегда будет указывать на действительный индекс в массиве images.

Во твоем варианте this.currentSlide сначала увеличивается на 1, а затем берется остаток от деления на длину массива изображений. Это может привести к проблемам, если this.currentSlide уже указывает на последний элемент массива, так как индекс последующего слайда может быть недопустимым.

Следовательно, как на меня то более предпочтительна моя, так как она более безопасна и гарантирует корректность индекса this.currentSlide.

Reply:
> Это может привести к проблемам, если this.currentSlide уже указывает на последний элемент массива, так как индекс последующего слайда может быть недопустимым.

Не совсем понимаю, в чем тут может быть опасность? Операция вычесления остатка на длину массива - как раз инструмент валидации выхода `currentIndex` за пределы массива; ежели нет, то я не понимаю, зачем он тут нужен)

Собственно, если мы рассмотрим пример с граничным значением, то получим:
```
const currentSLide = 8;
const productImagesLength = 9;

currentSLide += 1; // currentSLide === 9
currentSLide = currentSLide % productImagesLength; // currentSLide === 9 % 9 === 0
```
Из примера видно, что реализуется зацикливание `currentSLide`.

И по факту: код который я предложил и код который ты предложил - одно и то же), т.к. у тебя в синхронном коде тоже происходит присваивание инкрементированного не валидного значения. То есть в твоем примере в одной строке зашиты явные вычисления нового значения и неявная инициализация, я в своем - разнес по 2м разным действиям в рамках одной синхронной функции.

Вот через логи данного скрипта можно убедиться в правдивости высказывания выше:
```js
const arrayLength = 5;
let currentIndex = arrayLength - 1;

console.log((++currentIndex + (console.log('currentIndex in calculate', currentIndex) || 0)) % arrayLength);
```

Наверное повторюсь, но как по мне, всю эту логику нужно выносить в отдельную функцию чтобы не присваивать не валидные значения в целом и инкапсулировать все вычисления и валидации в одной отдельнолежащей функции

```ts
export function calculatingLoopedIndex(increment: -1 | 1, currentIndex: number, arrayLength: number): number {
    const newIndex = currentIndex + increment;
    const remainderOfDivision = newIndex % this.product.images.length;
    const isNegativeRemainderOfDivision =  remainderOfDivision < 0;
    
    return isNegativeRemainderOfDivision ? remainderOfDivision + arrayLength : remainderOfDivision;
}
```

```ts
```

Reply:
Кстати, а как ты планировал работу метода `prevSlide`? В моем понимании, сейчас он будет работать не корректно, т.к. выражение `--this.currentSLide % this.product.images.length` при значении `-1` будет возвращать `-1` 🤔

Reply:
Понравилось твое решение `calculatingLoopedIndex` и согласен с твоими комментариями.
**fix**

> выражение --this.currentSLide % this.product.images.length при значении -1 будет возвращать -1
А такого не должно быть - кнопку блокирую. Но согласен, что это не надёжно и лучше доп. проверка в методе
`<button mat-icon-button [disabled]="currentSLideIndex < 1" (click)="prevSlide()">
    <mat-icon>chevron_left</mat-icon>
</button>`

Reply:
> А такого не должно быть - кнопку блокирую. Но согласен, что это не надёжно и лучше доп. проверка в методе

Тут вопрос тогда в другом: если кнопка блокируется через UI, что в твоем случае является вполне валидным решением т.к. вся логика инкапсулирована в компоненте, тогда зачем идет деление на длинну массива, если это по сути функциональности не несет 🤔.  Т.е. в моем видении - либо делаем валидную проверку в обоих методах, либо не делаем ее вообще т.к. доступность вызова метода регулируется шаблоном компонента

Reply:
И снова ты прав!) - Замечание учту )